### DISCUSSION TITLE
Order of messages in EMQX 5.0

### DISCUSSION THREAD

**Discussion by sushiMix** [emqx/emqx#11144] [Q&A] [answered]
Hello, we have migrated from EMQX 4.x to EMQX 5.x and noticed a strange behavior with the new version.

When sending messages "quickly", we can observe an inversion of order when using mqttjs in web socket.

We first noticed it on subscription unsubscription of messages but it seems the same for messages on the same topic.

We have created a small test case in html which tries to send two consecutive publish message with different content and sometimes unsubscribe and subscribe again (the code send and receive the same messages).

The observed behavior is with EMQX 4.X there is no problem whereas with EMQX 5.0 an inversion is visible (and on first subscribe/subscribe no more messages are received).
Accordingly to our first analysis, it seems if MQTT packets are in the same websocket frame the order is not preserved.

We didn't see such behavior with paho mqtt clients (but they are no more maintained). We are still using mqtt 3.1.1 protocol version.

Did somebody else noticed it is there a tuning available to fix it ?

We tested on clustered an single node deployment

[emqx-order.html.zip](https://github.com/emqx/emqx/files/11880388/emqx-order.html.zip)

**Comment by lafirest** [comment]
@sushiMix 
Thanks for your feedback, we will look into this, but your attachment seems broken.

**Reply by lafirest** [reply]
~~We have just tried to reproduce your report, but failed, we think every things are ok here.
Could you please give us detailed reproduction steps, thanks~~
We have used the mqttS to successfully reproduce this.

**Comment by sushiMix** [comment]
Is it a real issue, shall I transform it to bug ?

**Reply by lafirest** [reply]
We also use other client SDKs for testing and can't reproduce this, we tentatively consider this a bug for mqttJS

**Reply by sushiMix** [reply]
We have noticed that other SDKs (like paho which is not maintained) do not send messages in the same web socket frame (and can't find in the specification that such behavior is forbidden).
>   A single WebSocket data frame can contain multiple or partial MQTT Control Packets. The receiver MUST NOT assume that MQTT Control Packets are aligned on WebSocket frame boundaries [MQTT-6.0.0-2].

When doing it (like mqttjs) we notice a difference of behavior between EMQX 4 and 5.
We see client send unsubscribe then subscribe in both cases (in same frame and for our logic in right order).
When EMQx 4 replies it is in the same order.
Whereas for EMQx 5 the acks are in reverse order.
The specification seems saying EMQx 4 had the right implementation :
- https://docs.oasis-open.org/mqtt/mqtt/v5.0/os/mqtt-v5.0-os.html#_Toc3901240
- http://docs.oasis-open.org/mqtt/mqtt/v3.1.1/os/mqtt-v3.1.1-os.html#_Toc398718105

The order send from client is still the right one (otherwise it would have failed for EMQx4). See attached PCAPs (tested on EMQX 4 and 5) for this use case.
[pcaps.tar.gz](https://github.com/emqx/emqx/files/11946327/pcaps.tar.gz).

**Reply by lafirest** [reply]
@sushiMix 
Sorry for this late reply, I know the order is reversed in EMQX 5, but from the debug log, the incoming packet id is right, seems the data are reversed before they have arrived at the broker.
May need to use the Wireshark for more deep investigation.

**Comment by lafirest** [comment]
@sushiMix 
I am so sorry for the late result of the investigation.
Yes, there is an order bug for packets in the same channel in version 5. x, I will fix it ASAP.

**Reply by lafirest** [reply]
the fix is [here](https://github.com/emqx/emqx/pull/11250)

**Reply by sushiMix** [reply]
Thanks, 
Do you know when the next docker build will be performed ?

**Reply by id** [reply, marked as answer]
@sushiMix the fix is included in https://github.com/emqx/emqx/releases/tag/v5.1.2

**Reply by sushiMix** [reply]
I successfully tested with EMQX 5.1.4 (but I have an issue with the persistent session and failover, I will check again and post in another thread)
